utils/skopeo.py

```diff
@@ -0,0 +1,239 @@
+import json
+import logging
+import re
+import subprocess
+
+from distutils import spawn
+
+
+class SkopeoCmdError(Exception):
+    """
+    Indicates that the skopeo command failed.
+    """
+
+
+class Skopeo:
+    """Wrapper around the skopeo utility."""
+
+    default_scheme = 'docker://'
+    default_repo = ''
+    default_tag = 'latest'
+
+    def __init__(self, dry_run=False):
+        self.dry_run = dry_run
+        self.log = logging.getLogger(__name__)
+        self.skopeo_cmd = spawn.find_executable('skopeo')
+        self._img_url_cache = dict()
+
+    def inspect(self, image):
+        """
+        Runs the skopeo "inspect" sub-command.
+
+        The skopeo "inspect" goes to the online repository and collects
+        multiple information about the image. That information is given in
+        JSON format, so we JSON-load it here before returning.
+
+        :param image: The image to be inspected.
+        :type image: str
+        :return: The JSON-loaded skopeo inspection information.
+        :rtype: dict
+        """
+        inspect_result = self._run_skopeo('inspect',
+                                          self._get_image_url(image))
+        try:
+            return json.loads(inspect_result)
+        except json.decoder.JSONDecodeError:
```

Comment:
if the json is invalid do we really want to be silent about it? I would let it raise an unhandled exception I think?

Reply:
The way this is implemented is:

- We have the quay destination, e.g. quay.io/app-sre/fedora
- We have the source image, e.g. docker.io/fedora:28
- We inspect docker.io/fedora:28, retrieving all the tags, copiyng them all.

The reason for this inspect to silently fail is that, when that happens, I still try to copy from docker.io/fedora:28 to quay.io/app-sre/fedora directly, without handling tags. Does that make sense?

Reply:
ok, sounds good